### PR TITLE
suggest `!` instead of erroneous `not` on if/while block parse failure

### LINKED ISSUE #46836: better syntax error for "not <condition>"  instead "!<condition>"
this code 
````rust
fn main() {
    funfunction(true);
}

fn funfunction(var: bool) {
    if not var  { println!("blubb");  }
}
````
currently emmits
````
  error: expected one of `,` or `}`, found `!`
 --> src/main.rs:6:26
  |
6 |     if not var  { println!("blubb");  }
  |                          ^ expected one of `,` or `}` here

error: expected `{`, found `var`
 --> src/main.rs:6:12
  |
6 |     if not var  { println!("blubb");  }
  |            ^^^-------------------------
  |            |
  |            help: try placing this code inside a block: `{ var{println,}; }`

error: aborting due to 2 previous errors

error: Could not compile `not`.

To learn more, run the command again with --verbose.

````

Could we detect explicitly if there are 2 unchained (via && or ||) "variables" after an "if" and if the first one is "not" and explicitly suggest using "!" instead of "not" instead?

meta:
````
rustc 1.24.0-nightly (dc39c3169 2017-12-17)
binary: rustc
commit-hash: dc39c31699a83313edf2ac096d0bf3cef871b705
commit-date: 2017-12-17
host: x86_64-unknown-linux-gnu
release: 1.24.0-nightly
LLVM version: 4.0
````

### PR DESCRIPTION
**Pull request by zackmdavis**
not

Impressing confused Python users with magical diagnostics is probably
worth this very slight (only 25ish lines) extra complexity in the
parser?

FIXME comments have been left to note that the formatting after
autofixing the suggestion (with `rustfix` or an RLS-powered IDE) won't
be optimal. (It doesn't look like any of the existing `CodeMap` methods
make it easy to adjust the span.)

Unfortunately, the parser doesn't seem smart enough to _know_ which of the "try placing this code inside a block" and `!` suggestions is correct. (Also, the function that sets the "inside a block" suggestion doesn't and shouldn't know about the condition, so canceling that suggestion if the `!` suggestion is issued would require a new clear-suggestions method on `Diagnostic`/`DiagnosticBuilder`, which I didn't want to write until we're convinced it's necessary.) The last function in the new UI test demonstrates a case in which the "block" suggestion is right and the `!` suggestion is wrong.

Resolves #46836.

r? @estebank

### CODE REVIEW

File: src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.stderr [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+LL |       if not::my department {
+   |       --         -^^^^^^^^^
+   |  _____|__________|
+   | |     |
+   | |     this `if` statement has a condition, but no block
+LL | |         //~^ ERROR expected `{`
+LL | |         println!("pass");
+LL | |         //~^ ERROR expected one of
+LL | |     }
+   | |_____- help: try placing this code inside a block: `{ department{println,}; }`
```
Comment (estebank) [inline src/test/ui/did_you_mean/issue-46836-identifier-not-instead-of-negation.stderr]: Can you also modify [`parse_block`](https://github.com/zackmdavis/rust/blob/b468b0e87c28e201427478bf6146afa0eaf85f81/src/libsyntax/parse/parser.rs#L4500) so that this suggestion is not given if the next token is an open brace, like in these cases? The suggestion as is is more likely to be wrong than right.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -3342,7 +3362,15 @@ impl<'a> Parser<'a> {
             return self.parse_while_let_expr(opt_label, span_lo, attrs);
         }
         let cond = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
-        let (iattrs, body) = self.parse_inner_attrs_and_block()?;
+        let (iattrs, body) = self.parse_inner_attrs_and_block()
+            .map_err(|mut err| {
+                if self.is_identifier_not(&cond) {
+                    let msg = "try replacing identifier `not` with the negation operator";
+                    // FIXME: replaced span doesn't include trailing whitespace
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: You could merge the `cond.span` with the next span, get the snippet and synthesize a span that _does_ include the whitespace (in a similar way as suggested in https://github.com/rust-lang/rust/pull/47574#issuecomment-367080392).

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -3232,6 +3247,11 @@ impl<'a> Parser<'a> {
         }
         let not_block = self.token != token::OpenDelim(token::Brace);
         let thn = self.parse_block().map_err(|mut err| {
+            if self.is_identifier_not(&cond) {
+                err.span_suggestion(cond.span, // FIXME: replaced span doesn't incl. whitespace
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: re: span, same as below

**Review by estebank — COMMENTED** [review]
LGTM, minor suggestions.

**Comment by bjorn3** [comment]
Could you please place the ``try replacing identifier `not` with the negation operator`` help message first, as I think it is the most likely solution when you use `not`.

**Comment by petrochenkov** [comment]
This is a quite limited ad hoc way to check for `not` used as `!`.
I'd expect it to work as normal parsing would work.
- You start parsing an expression (priority is important, should be the same as for `! EXPR`, i.e. `parse_prefix_expr`).
- If the current token is `not` and ...
- The next token is not in the "can continue expression after ident" set (you can look through `enum Token` and make this set, or I can do it tomorrow), then ...
- Report a non-fatal error, recover `not` as `!` and continue parsing.

This would cover tests from this PR and more.

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -3232,6 +3247,11 @@ impl<'a> Parser<'a> {
...
+                                    "try replacing identifier `not` with the negation operator",
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Looks a bit too wordy for a label, I'd personally use something "try replacing `not` with `!`", but as a minimum the word "identifier" can be removed here.

**Comment by petrochenkov** [comment]
>The next token is not in the "can continue expression after ident" set

So, the set (already inverted) is `Pound | Ident(..) | Literal(..) | NtIdent(..) | NtExpr(..) | NtBlock(..) | NtPath(..)` (this is a subset of `fn can_begin_expr`).

**Comment by zackmdavis** [comment]
> This is a quite limited ad hoc way to check for not used as !. I'd expect it to work as normal parsing would work.

Yeah, that's a good point.

> You start parsing an expression [...] So, the set (already inverted) is 

Thanks for the pointer! I can probably look into doing it this way later today or tomorrow.

File: src/libsyntax/parse/parser.rs (hunk 4 above)
Comment (zackmdavis) [reply src/libsyntax/parse/parser.rs]: The motivation was to emphasize that `not` has no special meaning in Rust-the-language (it's just another ordinary variable/identifier), even if `rustc`-the-compiler is (after this PR) nice enough to try to figure out notice what you meant, but I agree that _ceteris paribus_, longer messages are worse.

**Comment by zackmdavis** [comment]
I've [made some exciting progress](https://github.com/rust-lang/rust/commit/8c0f1f240e5ff84feb52e5978f47960fc48f7fb7) on the parse-prefix approach, but in addition to producing awesome diagnostics, my code also misparses some legitimate uses of `not` as a variable name. :disappointed: 

I'm pretty busy with a new dayjob lately, so I likely won't be able to look at this again for a week.

**Comment by petrochenkov** [comment]
@zackmdavis 
>also misparses some legitimate uses of not as a variable name

For example?
I certainly could miss something when making that token list.
Maybe something in macros?

**Comment by zackmdavis** [comment]
@petrochenkov 

> For example?

```
zmd@ReflectiveCoherence:~/Code/rust$ cat correct.rs 
fn main() {
    let not = 2;
    let _really = not;
}
zmd@ReflectiveCoherence:~/Code/rust$ rustc correct.rs # sane rustc
zmd@ReflectiveCoherence:~/Code/rust$ echo $?
0
zmd@ReflectiveCoherence:~/Code/rust$ rustc +stage1 correct.rs # my delusional work-in-progress build
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, or an operator, found `}`
 --> correct.rs:4:1
  |
3 |     let _really = not;
  |                      - expected one of 7 possible tokens here
4 | }
  | ^ unexpected token

error: aborting due to previous error
```

> I certainly could miss something when making that token list.

I think it's more likely to be an off-by-one-token error in my recovery logic, an error which we can pray will be revealed in the cold light of `info!` statements and careful thought. (But not tonight.)

(This commit is only using ident/literal/pound as the can't-continue-expression-after-ident set, because of the conjunction of my not feeling confident that I understand how the `Nt`s work, and their being a different enum which would require another `match` and the fact that leaving tokens out of the set should only decrease the number of situations in which we recover without affecting the correctness of the cases where we do try to recover.)

**Comment by petrochenkov** [comment]
>an off-by-one-token error

Ah, of course, the next token lookup is done with `look_ahead`, not `next_tok`.

**Comment by zackmdavis** [comment]
continued: #49258
